File: Interviews/GFG/Daily-Problems/2022/November/Day-08/Two-numbers-with-odd-occurrences.py
class Solution:
    def twoOddNum(self, Arr, N):
        xorr, bit, bucket1, bucket2 = 0, 37, 0, 0
        Arr = [int(i) for i in Arr]
        for i in Arr:
            xorr ^= i

        for i in range(37, -1, -1):
            if ((xorr >> i) & 1):
                bit = i
                break

        for i in Arr:
            if ((i >> bit) & 1):
                bucket1 ^= i
            else:
                bucket2 ^= i
        ans = [bucket1, bucket2]
        if bucket1 < bucket2:
            temp = ans[0]
            ans[0] = ans[-1]
            ans[-1] = temp
        return ans

File: Interviews/GFG/random-problems/Medium/Two-numbers-with-odd-occurrences/Two-numbers-with-odd-occurrences.cpp
class Solution {
   public:
    vector<long long int> twoOddNum(long long int Arr[], long long int N) {
        int64_t xorr = 0;
        for (int i = 0; i < N; i++) {
            xorr ^= Arr[i];
        }

        int bit = 37;
        for (int i = 37; i >= 0; i--) {
            if ((xorr >> i) & 1) {
                bit = i;
                break;
            }
        }

        int64_t bucket1 = 0, bucket2 = 0;
        for (int i = 0; i < N; i++) {
            if ((Arr[i] >> bit) & 1) {
                bucket1 ^= Arr[i];
            } else {
                bucket2 ^= Arr[i];
            }
        }

        vector<long long int> ans;
        if (bucket1 < bucket2)
            ans = {bucket2, bucket1};
        else
            ans = {bucket1, bucket2};

        return ans;
    }
};